updater/internal/state/monitor.go

```diff
@@ -0,0 +1,139 @@
+
+	// MonitorForSuccess will periodically check the state of the package. It will keep checking until the context is canceled or a failed/success state is detected.
+	// It will return an error if status is Failed or if the context times out.
+	MonitorForSuccess(ctx context.Context, packageName string) error
+}
+
+// CollectorMonitor implements Monitor interface for monitoring the Collector Package Status file
+type CollectorMonitor struct {
+	stateManager  packagestate.StateManager
+	currentStatus *protobufs.PackageStatuses
+}
+
+// NewCollectorMonitor create a new Monitor specifically for the collector
+func NewCollectorMonitor(logger *zap.Logger) (Monitor, error) {
+	// Get install directory
+	installDir, err := path.InstallDir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine install directory: %w", err)
+	}
+
+	// Create a collector monitor
+	packageStatusPath := filepath.Join(installDir, packagestate.DefaultFileName)
+	collectorMonitor := &CollectorMonitor{
+		stateManager: packagestate.NewFileStateManager(logger, packageStatusPath),
+	}
+
+	// Load the current status to ensure the package status file exists
+	collectorMonitor.currentStatus, err = collectorMonitor.stateManager.LoadStatuses()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load package statues: %w", err)
+	}
+
+	return collectorMonitor, nil
+
+}
+
+// SetState sets the status on the specified package and saves it to the package status file
+func (c *CollectorMonitor) SetState(packageName string, status protobufs.PackageStatus_Status, statusErr error) error {
+	// Verify we have package by that name
+	targetPackage, ok := c.currentStatus.GetPackages()[packageName]
+	if !ok {
+		return fmt.Errorf("no package for name %s", packageName)
+	}
+
+	// Update the status
+	targetPackage.Status = status
+
+	// If that passed in error is not nil set it as the error message
+	if statusErr != nil {
+		targetPackage.ErrorMessage = statusErr.Error()
+	}
+
+	c.currentStatus.GetPackages()[packageName] = targetPackage
+
+	// Save to updated status to disk
+	return c.stateManager.SaveStatuses(c.currentStatus)
+}
+
+// MonitorForSuccess intermittently checks the package status file for either an install failed or success status.
+// If an InstallFailed status is read this returns ErrFailedStatus error.
+// If the context is canceled or there is an error reading the status file the error will be returned.
+func (c *CollectorMonitor) MonitorForSuccess(ctx context.Context, packageName string) error {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			packageStatus, err := c.stateManager.LoadStatuses()
+			switch {
+			case errors.Is(err, os.ErrNotExist):
+				// File may not exist for a short period of time while being written so just continue
+				continue
+			case err != nil:
+				return fmt.Errorf("error while loaded status: %w", err)
```

Comment:
This is where I'm really paranoid; Some scenarios that we need to consider:

1. (This code already handles this one)
  * Collector deletes status file (it does this on every write IIRC)
  * Updater attempts to read status file
  * File does not exist, updater gets error

2. Read while write
  * Collector deletes status file, gets halfway through writing the status file
  * Updater attempts to read status file
  * Updater gets unmarshal error
3. Write while read
  * Updater opens file for reading
  * Updater partially reads file
  * Collector completely writes file
  * Updater reads whole file
  * This could be an unmarshal error, OR it could be completely valid depending on what was written, ending up being with the incorrect data being read.

I'm not sure if this is overly paranoid, but I think we should 
a. Basically ignore any error (probably log it), because unmarshal errors could happen, and IDK if we'll get any other weird errors contesting the file on windows.
b. Make sure the returned packages are static for two cycles, to make sure we've read what was actually supposed to be written, and not some partial of old and new data.

Those are my thoughts here, let me know if that sounds overly paranoid.

Reply:
@BinaryFissionGames i think this is all fair. 
i don't think the collector would ever write twice in a row during this process, so maybe just a single retry on error would be enough to handle 2. 
for 3. i think this would only happen if the collector detects a timeout on updater, and starts to modify status, but then the updater is trying to read at that point. this could be solved by simply making sure the collector kills the updater before writing in this case